src/address.c

```diff
@@ -0,0 +1,146 @@
+        return WALLY_EINVAL;
+
+    // Catch known incorrect combinations of address type and version:
+    if ((flags & WALLY_ADDRESS_TYPE_P2PKH && version == 0x05) ||
+        (flags & WALLY_ADDRESS_TYPE_P2SH_P2WPKH && version == 0x00) ||
+        (flags & WALLY_ADDRESS_TYPE_P2PKH && version == 0xC4) ||
+        (flags & WALLY_ADDRESS_TYPE_P2SH_P2WPKH && version == 0x6F))
+        return WALLY_EINVAL;
+
+    if (flags == WALLY_ADDRESS_TYPE_P2PKH) {
+        // pub_key_hash = ripemd160(sha256(pubkey)
+        address[0] = (unsigned char) version & 0xff;
+        if (wally_hash160(hdkey->pub_key, sizeof(hdkey->pub_key), address + 1, HASH160_LEN) != WALLY_OK)
+            return WALLY_EINVAL;
+    } else {
+        // redeem_script = SegWit version 0 + push(keyhash) = OP_0 + 0x20 + [key_hash]
+        // where key_hash = ripemd160(sha256(pubkey))
+        unsigned char redeem_script[HASH160_LEN + 2];
+        redeem_script[0] = OP_0;
+        redeem_script[1] = HASH160_LEN;
+
+        if (wally_hash160(hdkey->pub_key, sizeof(hdkey->pub_key), redeem_script + 2, HASH160_LEN) != WALLY_OK)
+            return WALLY_EINVAL;
+
+        // P2SH address = version (e.g. 0x05) + ripemd160(sha256(redeem_script))
+        address[0] = (unsigned char) version & 0xff;
+        if (wally_hash160(redeem_script, sizeof(redeem_script), address + 1, HASH160_LEN) != WALLY_OK)
+            return WALLY_EINVAL;
+    }
+
+    ret = wally_base58_from_bytes(address, sizeof(address), BASE58_FLAG_CHECKSUM, output);
+
+    wally_clear(address, sizeof(address));
+    return ret;
+}
+
+int wally_bip32_key_to_addr_segwit(const struct ext_key *hdkey, const char *addr_family,
+                                   uint32_t flags, char **output)
+{
+    int ret;
+
+    // Witness program bytes, including the version and data push opcode.
+    unsigned char witness_program_bytes[HASH160_LEN + 2];
+    witness_program_bytes[0] = OP_0;
+    witness_program_bytes[1] = HASH160_LEN;
+
+    if (wally_hash160(hdkey->pub_key, sizeof(hdkey->pub_key), witness_program_bytes + 2, HASH160_LEN) != WALLY_OK)
+        return WALLY_EINVAL;
+
+    ret = wally_addr_segwit_from_bytes(witness_program_bytes, HASH160_LEN + 2, addr_family, flags, output);
+    return ret;
+}
+
+int wally_address_to_scriptpubkey(const char *addr, uint32_t flags, unsigned char *bytes_out,
+                                  size_t len, size_t *written)
+{
+    int ret;
+
+    uint32_t version;
+    unsigned char bytes_base58_decode[1 + HASH160_LEN + BASE58_CHECKSUM_LEN];
+    size_t written_base58_decode;
+
+    if (written)
+        *written = 0;
+
+    if (!(flags == WALLY_NETWORK_BITCOIN_MAINNET || flags == WALLY_NETWORK_BITCOIN_TESTNET))
+        return WALLY_EINVAL;
+
+    // Check minimum length of bytes_out:
+    // P2PKH 25 bytes: OP_DUP OP_HASH160 [pub_key_hash] OP_EQUALVERIFY OP_CHECKSIG
+    // P2SH  23 bytes: OP_HASH160 [script_hash] OP_EQUAL
+    if (len < 25)
+        return WALLY_EINVAL;
```

Comment:
I think now we are calling wally_scriptpubkey_p2pkh_from_bytes/wally_scriptpubkey_p2sh_from_bytes we can drop this explicit length check because those functions themselves check len and return WALLY_EINVAL if len is too small for their respective outputs.

Reply:
Indeed, those functions alraedy check `len < WALLY_SCRIPTPUBKEY_P2PKH_LEN` and `_P2SH_LEN`.